File: estudos/programacao_orientada_a_objetos/java_oop/br/com/remember/oop/BrazilianCat.java
package br.com.remember.oop; //all classes in the same package to be mutually visible

//class definition: BrazilianCat --> inherits from (is a subclass of) Cat
public class BrazilianCat extends Cat {

    //doesn't need to declare any new attributes, inherits everything from the superclass

    //constructor function
    public BrazilianCat (String name) {
        super(name); //executes everything from the superclass constructor
    }

    @Override //this annotation indicates that this method exists in the superclass and is now being overwritten by the subclass
    public void talk () {
        //the attribute name is private to the Cat class, which means that not even its subclasses can access it freely... that's why we need to use the getName method here
        System.out.println(this.getName() + " says: miau!");
    }

    //this method is declared again with the same name, but with a different signature... it is now overloading the same method with a different behaviour, but doesn't overwrite the previous version
    public void talk (String message) {
        //the attribute name is private to the Cat class, which means that not even its subclasses can access it freely... that's why we need to use the getName method here
        System.out.println(this.getName() + " says: " + message);
    }
}

File: estudos/programacao_orientada_a_objetos/python_oop/BrazilianCat.py
from Cat import Cat #imports class Cat from file Cat.py (needed for inheritance)

#class definition: BrazilianCat --> inherits from (is a subclass of) Cat
class BrazilianCat (Cat) :

    #doesn't need to declare any new attributes, inherits everything from the superclass
    
    #constructor function
    def __init__ (self, name) :
        super().__init__(name) #executes everything from the superclass constructor
    
    #this method exists in the superclass and is now being overwritten by the subclass 
    def talk (self, message=None) :

        #the attribute __name is private to the Cat class, which means that not even its subclasses can access it freely... that's why we need to use the getName method here
        name = self.getName()

        #the behavior changes depending on the arguments passed when the method is called
        #behavior 1: when no argument is passed 
        if message is None :
            print("{} says: miau!".format(name))
        #behavior 2: when something is passed as argument
        else :
            print("{} says: {}".format(name, message))
        